### ISSUE TITLE
Devtool extensions don't show up on first load

### ISSUE THREAD

**Issue by MarshallOfSound** [electron/electron#6101]
- Electron version: `1.2.2+`
- Operating system: Windows

React Developer Tools work's as expected in `1.2.1` but any version above that reports a successful install but the Dev Tools do not appear.

I have tried both `file://` URL's and `http://` URL's, neither trigger the React DevTools.  This is because the `__REACT_DEVTOOLS_GLOBAL_HOOK__` global variable is never set.  Possibly an issue with running content scripts?

**Comment by kevinsawicki** [comment]
@MarshallOfSound what version of the react dev tools are you using?

I'm able to get them showing with Electron 1.2.3/React DevTools 0.14.11 when visiting http://todomvc.com/examples/react

Would you be able to provide a minimal sample app that shows them not working? I wonder if it might be related to some specific combination of options.

Also, are you trying to use them in a `<webview>` or just in a regular `BrowserWindow`?

<img width="1280" alt="screen shot 2016-06-17 at 9 02 12 am" src="https://cloud.githubusercontent.com/assets/671378/16156735/6ebf2cd8-346a-11e6-95a6-68f9762740ec.png">

**Comment by MarshallOfSound** [comment]
OK, so I discovered the source of my issue.

Asynchronously adding the dev tools extension does not make it load into the dev tools until you navigate to a new page, this behavior appeared to be inconsistent between `1.2.1` and `1.2.3` due to apparent load time differences between content scripts.

TLDR: Async loading a dev tools extension really doesn't work 👍

**Comment by MarshallOfSound** [comment]
In fact I'm tempted to re-open this as an issue.

Given the follow pseudo-code

```js
fetchExtension()
    .then(loadExtension)
    .then(() => {
        mainWindow = new BrowserWindow();
    });
```

Where `fetchExtension` downloads the extension and `loadExtension` calls the appropriate `BrowserWindow` methods.

On first load of `mainWindow` the React dev tools are not available.  On reload they become available.

Is this expected?

**Comment by MarshallOfSound** [comment]
[tmp-elec.zip](https://github.com/electron/electron/files/320968/tmp-elec.zip)

Minimum sample app is included

**Comment by kevinsawicki** [comment]
> Minimum sample app is included

Awesome, thanks 👍

**Comment by kevinsawicki** [comment]
@MarshallOfSound I'm seeing the following error when running your example:

```
An error occurred:  Chrome extensions can not be installed from the main process
```

Looks like it is logged from `electron-devtools-installer`.

**Comment by MarshallOfSound** [comment]
Oh, my bad I was using a locally linked dev version of that module.  I'll push and update that module so it works for you.  Gimme like 20 seconds :laughing:

**Comment by MarshallOfSound** [comment]
`electron-devtools-installer` has been updated and published, should work from the main process now.

**Comment by paulcbetts** [comment]
@MarshallOfSound Holy shit electron-devtools-installer, very cool

**Comment by MarshallOfSound** [comment]
@paulcbetts Yeah, I saw people including react dev tools and such inside their source code so they could include it easily.  I thought that was a super bad plan 😆

**Comment by MarshallOfSound** [comment]
@kevinsawicki Can we re-open this.  `1.2.4` has not fixed this issue

**Comment by kevinsawicki** [comment]
@MarshallOfSound

**Comment by kevinsawicki** [comment]
@MarshallOfSound sure, but I thought this was fixed, is your same example failing for you on 1.2.4?

**Comment by MarshallOfSound** [comment]
@kevinsawicki Yep, I just saw `1.2.4` was released so quickly ran the same sample against it.  React DevTools doesn't appear until you reload the page

**Comment by kevinsawicki** [comment]
@MarshallOfSound Hmm, I'm struggling to reproduce the issue using your example app on 1.2.4.

```js
app.on('ready', () => {
    installExtension('fmkadmapgofadopljbjfkapdkoienihi')
      .then((name) => {
        mainWindow = new (require('electron').BrowserWindow)();

        mainWindow.center();

        // and load the index.html of the app.
        mainWindow.loadURL(`http://todomvc.com/examples/react`);

        // Emitted when the window is closed.
        mainWindow.on('closed', () => {
          // Dereference the window object, usually you would store windows
          // in an array if your app supports multi windows, this is the time
          // when you should delete the corresponding element.
          mainWindow = null;
        });
      })
      .catch((err) => console.log('An error occurred: ', err));
  });
```

If I open the dev tools after the window shows then I see the React tab with content:

![screen shot 2016-06-22 at 4 03 47 pm](https://cloud.githubusercontent.com/assets/671378/16286688/eb07adc0-3892-11e6-9ebf-4f32f5cf09bb.png)

**Comment by MarshallOfSound** [comment]
@kevinsawicki Ah, it looks like I modified the example a tiny bit.

```js
app.on('ready', () => {
    installExtension('fmkadmapgofadopljbjfkapdkoienihi')
      .then((name) => {
        mainWindow = new (require('electron').BrowserWindow)();
        mainWindow.openDevTools();

        mainWindow.center();

        // and load the index.html of the app.
        mainWindow.loadURL(`http://todomvc.com/examples/react`);

        // Emitted when the window is closed.
        mainWindow.on('closed', () => {
          // Dereference the window object, usually you would store windows
          // in an array if your app supports multi windows, this is the time
          // when you should delete the corresponding element.
          mainWindow = null;
        });
      })
      .catch((err) => console.log('An error occurred: ', err));
  });
```

The key part is that `mainWindow.openDevTools();`, if I remove it and get back to the original sample you are correct.  The React tab appears perfectly, but if I open dev tools before navigation the React tab does not appear.

**Comment by codebytere** [comment]
@MarshallOfSound is this still happening? I'll close this otherwise

**Comment by MarshallOfSound** [comment]
@codebytere Considering 0 thumbs and I haven't been annoyed by it recently I'd say it doesn't happen any more 😆 Can always re-open if someone else reports it 👍
